### ISSUE TITLE
[JENKINS-73575] Jenkins master threads stuck on waiting

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-15#5171]
<p>Recently on my Jenkins I encountered a kind of thread lock, which turned out to be an indefinite wait which happened after restart with connection to one of the nodes, which was no longer existing, but jenkins master was waiting to connect to it anyway. Due to this wait, the master couldn't properly communicate with other nodes about ssh steps in the pipelines (none of the 'sh' steps in the pipelines worked). Additionally when trying to check logs from /logs/warning the endpoint was not responding. All the while the cpu, and memory load weren't high on the master instance.</p>

<p>The thread dump (2 blocked, waiting for non responding Channel@1a0969c9, which wasn't present in the master thread dump):<br/>
 <br/>
Channel reader thread: node_124<br/>
"Channel reader thread: node_124" Id=24000 Group=main WAITING on com.trilead.ssh2.channel.Channel@1a0969c9<br/>
at java.base@11.0.22/java.lang.Object.wait(Native Method)</p>
<ul class="alternate" type="square">
	<li> waiting on com.trilead.ssh2.channel.Channel@1a0969c9<br/>
at java.base@11.0.22/java.lang.Object.wait(Unknown Source)<br/>
at com.trilead.ssh2.channel.FifoBuffer.read(FifoBuffer.java:212)<br/>
at com.trilead.ssh2.channel.Channel$Output.read(Channel.java:127)<br/>
at com.trilead.ssh2.channel.ChannelManager.getChannelData(ChannelManager.java:935)<br/>
at com.trilead.ssh2.channel.ChannelInputStream.read(ChannelInputStream.java:58)<br/>
at com.trilead.ssh2.channel.ChannelInputStream.read(ChannelInputStream.java:79)<br/>
at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:94)<br/>
at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:74)<br/>
at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:105)<br/>
at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:39)<br/>
at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)<br/>
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:61)</li>
</ul>

<p>javamelody<br/>
"javamelody" Id=165 Group=main WAITING on com.trilead.ssh2.channel.Channel@1a0969c9<br/>
at java.base@11.0.22/java.lang.Object.wait(Native Method)</p>
<ul class="alternate" type="square">
	<li> waiting on com.trilead.ssh2.channel.Channel@1a0969c9<br/>
at java.base@11.0.22/java.lang.Object.wait(Unknown Source)<br/>
at com.trilead.ssh2.channel.ChannelManager.sendData(ChannelManager.java:385)<br/>
at com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:63)<br/>
at com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:68)<br/>
at hudson.remoting.ChunkedOutputStream.sendFrame(ChunkedOutputStream.java:93)<br/>
at hudson.remoting.ChunkedOutputStream.drain(ChunkedOutputStream.java:89)<br/>
at hudson.remoting.ChunkedOutputStream.write(ChunkedOutputStream.java:58)<br/>
at java.base@11.0.22/java.io.OutputStream.write(Unknown Source)<br/>
at hudson.remoting.ChunkedCommandTransport.writeBlock(ChunkedCommandTransport.java:45)<br/>
at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.write(AbstractSynchronousByteArrayCommandTransport.java:46)<br/>
at hudson.remoting.Channel.send(Channel.java:768)</li>
	<li> locked hudson.remoting.Channel@741652e2<br/>
at hudson.remoting.Request.callAsync(Request.java:238)<br/>
at hudson.remoting.Channel.callAsync(Channel.java:1032)<br/>
at net.bull.javamelody.RemoteCallHelper.collectDataByNodeName(RemoteCallHelper.java:189)<br/>
at net.bull.javamelody.RemoteCallHelper.collectJavaInformationsListByName(RemoteCallHelper.java:214)<br/>
at net.bull.javamelody.NodesCollector.collectWithoutErrorsNow(NodesCollector.java:159)<br/>
at net.bull.javamelody.NodesCollector.collectWithoutErrors(NodesCollector.java:147)<br/>
at net.bull.javamelody.NodesCollector$2.run(NodesCollector.java:115)<br/>
at java.base@11.0.22/java.util.TimerThread.mainLoop(Unknown Source)<br/>
at java.base@11.0.22/java.util.TimerThread.run(Unknown Source)<br/>
 <br/>
The issue looks like a rare race condition when the agent is deleted midway communication, causing the master to be stuck on 'wait' method.<br/>
 <br/>
Proposed solution:<br/>
Could I/you update the parts:<br/>
<a href="https://github.com/jenkinsci/trilead-ssh2/blob/main/src/com/trilead/ssh2/channel/FifoBuffer.java#L212" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/trilead-ssh2/blob/main/src/com/trilead/ssh2/channel/FifoBuffer.java#L212</a><br/>
and<br/>
<a href="https://github.com/jenkinsci/trilead-ssh2/blob/main/src/com/trilead/ssh2/channel/ChannelManager.java#L385" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/trilead-ssh2/blob/main/src/com/trilead/ssh2/channel/ChannelManager.java#L385</a><br/>
so that it will be: 'wait(900000)' 15min ?<br/>
Or create a new variable like: DEFAULT_CONNECTION_TIMEOUT_SECONDS, or use any other existing one, as long as it will have some timeout eventually.<br/>
 <br/>
There's total 8 uses of 'wait' function, so it'd be good to update all of them.</li>
</ul>

<p>Unfortunately I wasn't able to reproduce the error so far as it's very time sensitive bug to produce.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=experrior">experrior</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-73575" target="_blank">Jenkins master threads stuck on waiting</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=experrior">experrior</a>
<li><b>status</b>: In Progress
<li><b>priority</b>: Minor
<li><b>component(s)</b>: core
<li><b>label(s)</b>: core, deadlock, trilead-ssh2-jenkins
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-11-16
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Recently on my Jenkins I encountered a kind of thread lock, which turned out to be an indefinite wait which happened after restart with connection to one of the nodes, which was no longer existing, but jenkins master was waiting to connect to it anyway. Due to this wait, the master couldn't properly communicate with other nodes about ssh steps in the pipelines (none of the 'sh' steps in the pipelines worked). Additionally when trying to check logs from /logs/warning the endpoint was not responding. All the while the cpu, and memory load weren't high on the master instance.</p>

<p>The thread dump (2 blocked, waiting for non responding Channel@1a0969c9, which wasn't present in the master thread dump):
 
Channel reader thread: node_124
"Channel reader thread: node_124" Id=24000 Group=main WAITING on com.trilead.ssh2.channel.Channel@1a0969c9
at java.base@11.0.22/java.lang.Object.wait(Native Method)</p>
<ul class="alternate" type="square">
	<li> waiting on com.trilead.ssh2.channel.Channel@1a0969c9
at java.base@11.0.22/java.lang.Object.wait(Unknown Source)
at com.trilead.ssh2.channel.FifoBuffer.read(FifoBuffer.java:212)
at com.trilead.ssh2.channel.Channel$Output.read(Channel.java:127)
at com.trilead.ssh2.channel.ChannelManager.getChannelData(ChannelManager.java:935)
at com.trilead.ssh2.channel.ChannelInputStream.read(ChannelInputStream.java:58)
at com.trilead.ssh2.channel.ChannelInputStream.read(ChannelInputStream.java:79)
at hudson.remoting.FlightRecorderInputStream.read(FlightRecorderInputStream.java:94)
at hudson.remoting.ChunkedInputStream.readHeader(ChunkedInputStream.java:74)
at hudson.remoting.ChunkedInputStream.readUntilBreak(ChunkedInputStream.java:105)
at hudson.remoting.ChunkedCommandTransport.readBlock(ChunkedCommandTransport.java:39)
at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:34)
at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:61)</li>
</ul>

<p>javamelody
"javamelody" Id=165 Group=main WAITING on com.trilead.ssh2.channel.Channel@1a0969c9
at java.base@11.0.22/java.lang.Object.wait(Native Method)</p>
<ul class="alternate" type="square">
	<li> waiting on com.trilead.ssh2.channel.Channel@1a0969c9
at java.base@11.0.22/java.lang.Object.wait(Unknown Source)
at com.trilead.ssh2.channel.ChannelManager.sendData(ChannelManager.java:385)
at com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:63)
at com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:68)
at hudson.remoting.ChunkedOutputStream.sendFrame(ChunkedOutputStream.java:93)
at hudson.remoting.ChunkedOutputStream.drain(ChunkedOutputStream.java:89)
at hudson.remoting.ChunkedOutputStream.write(ChunkedOutputStream.java:58)
at java.base@11.0.22/java.io.OutputStream.write(Unknown Source)
at hudson.remoting.ChunkedCommandTransport.writeBlock(ChunkedCommandTransport.java:45)
at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.write(AbstractSynchronousByteArrayCommandTransport.java:46)
at hudson.remoting.Channel.send(Channel.java:768)</li>
	<li> locked hudson.remoting.Channel@741652e2
at hudson.remoting.Request.callAsync(Request.java:238)
at hudson.remoting.Channel.callAsync(Channel.java:1032)
at net.bull.javamelody.RemoteCallHelper.collectDataByNodeName(RemoteCallHelper.java:189)
at net.bull.javamelody.RemoteCallHelper.collectJavaInformationsListByName(RemoteCallHelper.java:214)
at net.bull.javamelody.NodesCollector.collectWithoutErrorsNow(NodesCollector.java:159)
at net.bull.javamelody.NodesCollector.collectWithoutErrors(NodesCollector.java:147)
at net.bull.javamelody.NodesCollector$2.run(NodesCollector.java:115)
at java.base@11.0.22/java.util.TimerThread.mainLoop(Unknown Source)
at java.base@11.0.22/java.util.TimerThread.run(Unknown Source)
 
The issue looks like a rare race condition when the agent is deleted midway communication, causing the master to be stuck on 'wait' method.
 
Proposed solution:
Could I/you update the parts:
<a href="https://github.com/jenkinsci/trilead-ssh2/blob/main/src/com/trilead/ssh2/channel/FifoBuffer.java#L212" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/trilead-ssh2/blob/main/src/com/trilead/ssh2/channel/FifoBuffer.java#L212</a>
and
<a href="https://github.com/jenkinsci/trilead-ssh2/blob/main/src/com/trilead/ssh2/channel/ChannelManager.java#L385" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/trilead-ssh2/blob/main/src/com/trilead/ssh2/channel/ChannelManager.java#L385</a>
so that it will be: 'wait(900000)' 15min ?
Or create a new variable like: DEFAULT_CONNECTION_TIMEOUT_SECONDS, or use any other existing one, as long as it will have some timeout eventually.
 
There's total 8 uses of 'wait' function, so it'd be good to update all of them.</li>
</ul>

<p>Unfortunately I wasn't able to reproduce the error so far as it's very time sensitive bug to produce.</p></pre>
</details>
</details>
<details><summary><i>1 attachment</i></summary>

- [image-2024-08-19-15-16-48-527.png](https://issues.jenkins.io/secure/attachment/63169/image-2024-08-19-15-16-48-527.png)
  > ![image-2024-08-19-15-16-48-527.png](https://issues.jenkins.io/secure/attachment/63169/image-2024-08-19-15-16-48-527.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>experrior's <a href="https://issues.jenkins.io/browse/JENKINS-73575?focusedId=448844&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-448844">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>I've recently identified another issue with com.trilead.ssh2:</p>

<p>On the long running Jenkins with many agents, there was a problem with monitoring plugin, it stopped recording graphs for quite long time:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/63169/63169_image-2024-08-19-15-16-48-527.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p> </p>

<p>Waiting thread:</p>

<p>"javamelody" daemon prio=5 WAITING
    java.lang.Object.wait(Native Method)
    java.lang.Object.wait(Object.java:502)
    com.trilead.ssh2.channel.ChannelManager.sendData(ChannelManager.java:383)
    com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:63)
    com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:68)
    hudson.remoting.ChunkedOutputStream.sendFrame(ChunkedOutputStream.java:89)
    hudson.remoting.ChunkedOutputStream.drain(ChunkedOutputStream.java:85)
    hudson.remoting.ChunkedOutputStream.write(ChunkedOutputStream.java:54)
    java.io.OutputStream.write(OutputStream.java:75)
    hudson.remoting.ChunkedCommandTransport.writeBlock(ChunkedCommandTransport.java:45)
    hudson.remoting.AbstractSynchronousByteArrayCommandTransport.write(AbstractSynchronousByteArrayCommandTransport.java:46)
    hudson.remoting.Channel.send(Channel.java:764)
    hudson.remoting.Request.callAsync(Request.java:238)
    hudson.remoting.Channel.callAsync(Channel.java:1028)
    net.bull.javamelody.RemoteCallHelper.collectDataByNodeName(RemoteCallHelper.java:188)
    net.bull.javamelody.RemoteCallHelper.collectJavaInformationsListByName(RemoteCallHelper.java:213)
    net.bull.javamelody.NodesCollector.collectWithoutErrorsNow(NodesCollector.java:159)
    net.bull.javamelody.NodesCollector.collectWithoutErrors(NodesCollector.java:147)
    net.bull.javamelody.NodesCollector$2.run(NodesCollector.java:115)
    java.util.TimerThread.mainLoop(Timer.java:555)
    java.util.TimerThread.run(Timer.java:505)</p>



<p>Again forever stuck by not having timeout in the method.</p>
</pre>
</details>

</details>
<p>I've recently identified another issue with com.trilead.ssh2:</p>

<p>On the long running Jenkins with many agents, there was a problem with monitoring plugin, it stopped recording graphs for quite long time:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/63169/63169_image-2024-08-19-15-16-48-527.png" style="border: 0px solid black" /></span></p>

<p> </p>

<p> </p>

<p>Waiting thread:</p>

<p>"javamelody" daemon prio=5 WAITING<br/>
    java.lang.Object.wait(Native Method)<br/>
    java.lang.Object.wait(Object.java:502)<br/>
    com.trilead.ssh2.channel.ChannelManager.sendData(ChannelManager.java:383)<br/>
    com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:63)<br/>
    com.trilead.ssh2.channel.ChannelOutputStream.write(ChannelOutputStream.java:68)<br/>
    hudson.remoting.ChunkedOutputStream.sendFrame(ChunkedOutputStream.java:89)<br/>
    hudson.remoting.ChunkedOutputStream.drain(ChunkedOutputStream.java:85)<br/>
    hudson.remoting.ChunkedOutputStream.write(ChunkedOutputStream.java:54)<br/>
    java.io.OutputStream.write(OutputStream.java:75)<br/>
    hudson.remoting.ChunkedCommandTransport.writeBlock(ChunkedCommandTransport.java:45)<br/>
    hudson.remoting.AbstractSynchronousByteArrayCommandTransport.write(AbstractSynchronousByteArrayCommandTransport.java:46)<br/>
    hudson.remoting.Channel.send(Channel.java:764)<br/>
    hudson.remoting.Request.callAsync(Request.java:238)<br/>
    hudson.remoting.Channel.callAsync(Channel.java:1028)<br/>
    net.bull.javamelody.RemoteCallHelper.collectDataByNodeName(RemoteCallHelper.java:188)<br/>
    net.bull.javamelody.RemoteCallHelper.collectJavaInformationsListByName(RemoteCallHelper.java:213)<br/>
    net.bull.javamelody.NodesCollector.collectWithoutErrorsNow(NodesCollector.java:159)<br/>
    net.bull.javamelody.NodesCollector.collectWithoutErrors(NodesCollector.java:147)<br/>
    net.bull.javamelody.NodesCollector$2.run(NodesCollector.java:115)<br/>
    java.util.TimerThread.mainLoop(Timer.java:555)<br/>
    java.util.TimerThread.run(Timer.java:505)</p>



<p>Again forever stuck by not having timeout in the method.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ifernandezcalvo.png" title="ifernandezcalvo's avatar" /> ifernandezcalvo's <a href="https://issues.jenkins.io/browse/JENKINS-73575?focusedId=448895&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-448895">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>If you configure the TCP stack of your Jenkins Controller and Jenkins Agents to a proper value for TIME_WAIT, The connection will die and the threads. The default value in Linux is 7200 seconds, I usually recommend to set it to 120 seconds.</p>
</pre>
</details>

</details>
<p>If you configure the TCP stack of your Jenkins Controller and Jenkins Agents to a proper value for TIME_WAIT, The connection will die and the threads. The default value in Linux is 7200 seconds, I usually recommend to set it to 120 seconds.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ifernandezcalvo.png" title="ifernandezcalvo's avatar" /> ifernandezcalvo's <a href="https://issues.jenkins.io/browse/JENKINS-73575?focusedId=448964&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-448964">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Again, you have an issue of SSH connections half closed, that ends on a tons of thread block forever. The solution is to configure the TCP stack to end this connections. The change on the wait you propose in the PR could help but the half closed connection will still there if you do not configure your TCP stack.</p>
</pre>
</details>

</details>
<p>Again, you have an issue of SSH connections half closed, that ends on a tons of thread block forever. The solution is to configure the TCP stack to end this connections. The change on the wait you propose in the PR could help but the half closed connection will still there if you do not configure your TCP stack.</p>
